### ISSUE TITLE
General questions about pytorch example source code

### ISSUE THREAD

**Issue by erichu2000** [horovod/horovod#2191]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Hi!
I'm new to horovod and have general questions regarding some of the basic concepts in the restnet50 source code.
Question1: What does rank mean in horovod? For instance, "Gradient is applied across all ranks" what does the word rank mean in this comment? What does rank mean in hvd.rank()?

Question2: train_loader = torch.utils.data.Dataloader(blablabla), what specifically does train_loader now equals to? 

Question3:

**Comment by tgaddair** [comment]
Hi @erichu2000, to answer your questions:

1. Rank is an MPI concept that essentially means "worker ID".  If you train with `horovodrun -np 4`, you have 4 processes with ranks `[0, 1, 2, 3]`. We provide more details about these concepts [here](https://github.com/horovod/horovod/blob/master/docs/concepts.rst).

2. Not sure I understand your second question, but the `train_loader` will be a PyTorch Dataloader that only operates on a subset of the data each epoch.  The DistributedSampler's job is to partition the dataset among the workers. When the Dataloader selects the next sample, it gets it from the DistributedSampler.

**Comment by erichu2000** [comment]
@tgaddair Thanks a lot! Very clear explanation. We are trying to build a system that automatically converts user input(a machine learning python file, for instance SSD or YOLO, that runs only a single machine) to a horovod based model that could run on multiple machines. Now we're using regular expression as a way to accomplish this goal, however, regular expression brings a lot of problem when it comes to the user uses a lot of customized functions and classes, etc. Are there any suggestions for the best approach to this problem? Perhaps using an interpreter? Thanks in advance！

**Comment by tgaddair** [comment]
Hey @erichu2000, if I understand the issue correctly, that you want to parse user input on a single process, then spawn a Horovod distributed job, then one way to accomplish that is to use the `horovod.run` API, which allows you to launch the job programmatically (note that in the released version, it's under `horovod.run.run`.  If there's anything else you need clarity on, please feel free to share some code snippets and I'll take a look.
